platform/api.progress.nb/src/org/netbeans/api/progress/ProgressHandleFactory.java

```diff
@@ -119,44 +148,171 @@ public static JLabel createMainLabelComponent(ProgressHandle handle) {
      *                   to an output of the task. The action is assumed to open the apropriate component with the task's output.
      * @param displayName to be shown in the progress UI
      * @return an instance of {@link org.netbeans.api.progress.ProgressHandle}, initialized but not started.
-     *
+     * @deprecated Please use {@link ProgressHandle#createSystemHandle(java.lang.String, org.openide.util.Cancellable, javax.swing.Action)}
+     * @Deprecated
      */
     public static ProgressHandle createSystemHandle(String displayName, Cancellable allowToCancel, Action linkOutput) {
+        return ProgressHandle.createSystemHandle(displayName, allowToCancel, linkOutput);
+    }    
+
+    /**
+     * Create a progress UI handle for a task that is not triggered by explicit user action.
+     * This call creates a Swing-based implementation. Use only when the Handle is directly used in
+     * Swing UIs. These handles will never work in other presenters.
+     * @param allowToCancel either null, if the task cannot be cancelled or 
+     *          an instance of {@link org.openide.util.Cancellable} that will be called when user 
+     *          triggers cancel of the task.
+     * @param linkOutput an <code>Action</code> instance that links the running task in the progress bar
+     *                   to an output of the task. The action is assumed to open the apropriate component with the task's output.
+     * @param displayName to be shown in the progress UI
+     * @return an instance of {@link org.netbeans.api.progress.ProgressHandle}, initialized but not started.
+     * @since 1.59
+     */
+    public static ProgressHandle createSystemUIHandle(String displayName, Cancellable allowToCancel, Action linkOutput) {
         return new UIInternalHandle(displayName, allowToCancel, false, linkOutput).createProgressHandle();
     }    
+
+    private static ProgressUIWorkerProvider TRIVIAL_PROVIDER = new TrivialProgressUIWorkerProvider();
+    
+    private static ExtractedProgressUIWorker ihextract(ProgressHandle h) {
+        InternalHandle ih = h.getInternalHandle();
+        if (ih instanceof UIInternalHandle) {
+            return new UIHandleExtractor((UIInternalHandle)ih);
+        } else {
+            // fallback for non-UIInternalHandles. Let the environment to create
+            // a suitable component representation.
+            ProgressUIWorkerProvider prov = Lookup.getDefault().lookup(ProgressUIWorkerProvider.class);
+            if (prov == null) {
+                prov = TRIVIAL_PROVIDER;
+            }
+            ExtractedProgressUIWorker worker = prov.extractProgressWorker(ih);
+            if (worker != null) {
+                return new StrangeExtractor(ih, worker);
+            } else {
+                return null;
+            }
+        }
+    }
+    
+    private static class StrangeExtractor implements ExtractedProgressUIWorker {
```

Comment:
`StrangeExtractor` is really a strange name! NonUI, Foreign, Generic, Delegating?